Lib/fontParts/base/base.py

```diff
@@ -236,125 +307,267 @@ def raiseNotImplementedError(self):
 
-            >>> obj.changed()
-        """
+        The behavior of this method will vary from environment to environment.
+
+        :param \*args: Any positional arguments.
+        :param \**kwargs: Any keyword arguments.
 
-    def naked(self):
         """
-        Return the environment's native object
-        that has been wrapped by this object.
 
-            >>> loweLevelObj = obj.naked()
+    def naked(self) -> Any:
+        """Return the environment's native object wrapped by the current object.
+
+        :raises NotImplementedError: If the method has not been overridden by a
+            subclass.
+
         """
         self.raiseNotImplementedError()
 
 
 class BaseDict(BaseObject):
+    """Provide objects with basic dictionary-like functionality.
+
+    :cvar keyNormalizer: An optional normalizer function for keys.
+    :cvar valueNormalizer: An optional normalizer function for values.
+
+    """
+    keyNormalizer: Optional[Any] = None
+    valueNormalizer: Optional[Any] = None
+
+    def copyData(self, source: BaseDict) -> None:
+        """Copy data from another object instance.
 
-    keyNormalizer = None
-    valueNormalizer = None
+        This method calls the superclass's `copyData` method and updates
+        the current dictionary with the contents of the `source`.
 
-    def copyData(self, source):
+        param source: The source :class:`BaseDict` instance from which
+            to copy data.
+
+        """
         super(BaseDict, self).copyData(source)
         self.update(source)
 
-    def __len__(self):
+    def __len__(self) -> int:
+        """Return the number of items in the object.
+
+        :return: An :class:`int` representing the number of dictionary items.
+
+        """
         value = self._len()
         return value
 
-    def _len(self):
-        """
-        Subclasses may override this method.
+    def _len(self) -> int:
+        """Return the number of items in the native object.
+
+        This is the environment implementation of :attr:`BaseDict.__len__`.
+
+        :return: An :class:`int` representing the number of dictionary items.
+
+        .. note::
+
+            Subclasses may override this method.
+
         """
         return len(self.keys())
 
-    def keys(self):
+    def keys(self) -> List[Any]:
+        """Return a list of keys in the object.
+
+        :return: A :class:`list` of dictionary keys.
+
+        """
         keys = self._keys()
         if self.keyNormalizer is not None:
             keys = [self.keyNormalizer.__func__(key) for key in keys]
         return keys
 
-    def _keys(self):
-        """
-        Subclasses may override this method.
+    def _keys(self) -> List[Any]:
+        """Return a list of keys in the native object.
+
+        This is the environment implementation of :meth:`BaseDict.keys`.
+
+        :return: A :class:`list` of dictionary keys. If
+            a :cvar:`BaseDict.keyNormalizer` is set, it will have been applied
```

Comment:
better to say `it will be applied to each key in the calling method`

Reply:
Changed.